metriq_gym/benchmarks/quantum_volume.py

```diff
@@ -1,45 +1,207 @@
+    circuits = []
+    ideal_probs = []
+
+    for _ in range(trials):
+        circuit = qiskit_random_circuit_sampling(n)
+        sim_circuit = circuit.copy()
+        circuit.measure_all()
+        circuits.append(circuit)
+        sim = QrackSimulator(n)
+        sim.run_qiskit_circuit(sim_circuit, shots=0)
+        ideal_probs.append(sim.out_probs())
+        del sim
+
+    return circuits, ideal_probs
+
+
+@dataclass
+class TrialStats:
+    """Data class to store statistics of a single trial.
+
+    Attributes:
+        qubits: Number of qubits used in the circuit.
+        shots: Number of measurement shots performed on the quantum circuit.
+        xeb: Cross Entropy Benchmarking score.
+        hog_prob: Probability of measuring heavy outputs.
+        hog_pass: Boolean indicating whether the heavy output probability exceeds 2/3.
+        p_value: p-value for the heavy output count.
+        confidence_level: Confidence level for benchmarking.
+        confidence_pass: Boolean indicating if the p-value is below the confidence level.
+        eplg: Estimated Pauli Layer Gate (EPLG) fidelity.
+    """
+
+    qubits: int
+    shots: int
+    xeb: float
+    hog_prob: float
+    hog_pass: bool
+    p_value: float
+    confidence_level: float
+    confidence_pass: bool
+    eplg: float
+
+
+@dataclass
+class AggregateStats:
+    """Data class to store aggregated statistics over multiple trials.
+
+    Attributes:
+        provider: The quantum backend provider for the result.
+        trials: Number of trials aggregated.
+        trial_p_values: List of p-values for each trial.
+        hog_prob: Average probability of measuring heavy outputs across trials.
+        p_value: Combined p-value for all trials.
+        hog_pass: Boolean indicating whether all trials exceeded the heavy output probability threshold.
+        confidence_pass: Boolean indicating if all trials passed the confidence level.
+    """
+
+    trials: int
+    trial_p_values: list[float]
+    hog_prob: float
+    p_value: float
+    hog_pass: bool
+    confidence_pass: bool
+
 
-        logging.info(f"Dispatched {self.params["trials"]} trials in 1 job.")
+def calc_trial_stats(
+    ideal_probs: list[float],
+    counts: dict[str, int],
+    shots: int,
+    confidence_level: float,
+) -> TrialStats:
+    """Calculate various statistics for quantum volume benchmarking.
 
-        self.job_manager.add_job(result.to_serializable())
+    Args:
+        ideal_probs: A dictionary of bitstrings to ideal probabilities.
+        counts: A dictionary of bitstrings to counts measured from the backend.
+        shots: Number of measurement shots performed on the quantum circuit.
+        confidence_level: Specified confidence level for the benchmarking.
 
-        logging.info("Dispatched job saved.")
+    Returns:
+        A `TrialStats` object containing the calculated statistics.
+    """
+    n_pow = len(ideal_probs)
+    n = int(round(math.log2(n_pow)))
```

Comment:
[question]: If this number is being calculated here, why do we directly put in a number for `num_qubits` when specifying the quantum volume schema? It seems to be overwritten here.